File: packages/nodejs/src/domains/auth/email.ts
import { AUTH_TYPE } from '../../constants';
import { verifyPassword } from '../../helpers';
import { signinFailed } from '../../errors';
import { findOneWithCredentialByEmail } from '../adminuser';
import { createFirstAdminUser } from './common';
import { signJwt } from './jwt';
import http from 'http';

export interface EmailConfig {
  jwt: {
    issuer: string;
    audience: string;
  };
}

// Emailアドレスでサインイン
export const signinEmail = async (
  req: http.IncomingMessage,
  email: string,
  password: string
): Promise<string> => {
  // credentialありで取得
  let adminUser = await findOneWithCredentialByEmail(email);
  if (!adminUser) {
    adminUser = await createFirstAdminUser(AUTH_TYPE.EMAIL, {
      email,
      password,
    });
    if (!adminUser) {
      // 他に管理者がいる場合は発行してもらう必要がある
      console.error(
        'No admin user found. Please contact the administrator to create an account.'
      );
      throw signinFailed();
    }
  }

  if (
    !adminUser.password ||
    !adminUser.salt ||
    !verifyPassword(password, adminUser.password, adminUser.salt)
  ) {
    console.error('Invalid password');
    throw signinFailed();
  }

  return signJwt(adminUser.id, req);
};

File: packages/golang/domains/auth/email.go
package auth

import (
	"net/http"

	"github.com/cam-inc/viron/packages/golang/helpers"

	"github.com/cam-inc/viron/packages/golang/errors"

	"github.com/cam-inc/viron/packages/golang/constant"
	"github.com/cam-inc/viron/packages/golang/domains"
)

// SigninEmail Emailアドレスでサインイン
func SigninEmail(r *http.Request, email string, password string) (string, *errors.VironError) {
	ctx := r.Context()
	user := domains.FindByEmail(ctx, email)
	log.Debugf("user(%s)", email)
	if user == nil {
		log.Debugf("user(%s) is nil", email)
		payload := &domains.AdminUser{
			Email:    email,
			Password: &password,
		}
		var err error
		user, _, err = createFirstAdminUser(ctx, payload, nil, constant.AUTH_TYPE_EMAIL)
		if err != nil || user == nil {
			return "", errors.SigninFailed
		}
	}

	if !helpers.VerifyPassword(password, *user.Password, *user.Salt) {
		return "", errors.SigninFailed
	}

	token, err := Sign(r, user.ID)
	if err != nil {
		log.Error("SigninEmail sign failed %#v \n", err)
		return "", errors.SigninFailed
	}
	return token, nil
}
